src/ess/amor/utils.py

```diff
@@ -70,49 +74,189 @@ def theta_grid(
-    da = da.bins.concat(set(da.dims) - {"wavelength"}).hist(theta=thbins).transpose()
-    p = da.plot(norm="log")
-    for a in sc.linspace("_", 0.1, 3, 10):
+def wavelength_theta_figure(
+    da: sc.DataArray | Sequence[sc.DataArray],
+    *,
+    wavelength_bins: (sc.Variable | None) | Sequence[sc.Variable | None] = None,
+    theta_bins: (sc.Variable | None) | Sequence[sc.Variable | None] = None,
+    q_edges_to_display: Sequence[sc.Variable] = (),
+    linewidth: float = 1.0,
+    **kwargs,
+):
+    if isinstance(da, sc.DataArray):
+        return wavelength_theta_figure(
+            (da,),
+            wavelength_bins=(wavelength_bins,),
+            theta_bins=(theta_bins,),
+            q_edges_to_display=q_edges_to_display,
+            **kwargs,
+        )
+
+    wavelength_bins, theta_bins = (
+        (None,) * len(da)
+        if v is None
+        else (v,) * len(da)
+        if isinstance(v, sc.Variable)
+        else v
+        for v in (wavelength_bins, theta_bins)
+    )
+
+    hs = []
+    for da, wavelength_bins, theta_bins in zip(  # noqa: B020
+        da, wavelength_bins, theta_bins, strict=True
+    ):
+        if da.bins:
+            da = da.bins.concat(set(da.dims) - {"wavelength", "theta"})
+        all_coords = {*da.coords, *(da.bins or da).coords}
+        if 'wavelength' not in all_coords or 'theta' not in all_coords:
+            raise ValueError('Data must have wavelength and theta coord')
+        if da.bins or set(da.dims) != {"wavelength", "theta"}:
+            bins = {}
+            if 'sample_rotation' in da.coords and 'detector_rotation' in da.coords:
+                bins['theta'] = theta_grid(
+                    nu=da.coords['detector_rotation'], mu=da.coords['sample_rotation']
+                )
+            if theta_bins is not None:
+                bins['theta'] = theta_bins
+            if wavelength_bins is not None:
+                bins['wavelength'] = wavelength_bins
+            if 'theta' not in da.dims and 'theta' not in bins:
+                raise ValueError('No theta binning provided')
+            if 'wavelength' not in da.dims and 'wavelength' not in bins:
+                raise ValueError('No wavelength binning provided')
+            da = da.hist(**bins)
+
+        hs.append(da.transpose(('theta', 'wavelength')))
+
+    kwargs.setdefault('cbar', True)
+    kwargs.setdefault('norm', 'log')
+    p = pp.imagefigure(*(pp.Node(h) for h in hs), **kwargs)
+    for q in q_edges_to_display:
+        thmax = max(h.coords["theta"].max() for h in hs)
         p.ax.plot(
-            [sc.scalar(0.0), da.coords["wavelength"].max().value],
-            [sc.scalar(0.0), a * da.coords["theta"].max().value],
+            [0.0, 4 * np.pi * (sc.sin(thmax) / q).value],
+            [0.0, thmax.value],
             linestyle="solid",
-            linewidth=0.5,
+            linewidth=linewidth,
             color="black",
             marker=None,
         )
     return p
 
 
+def q_theta_figure(
```

Comment:
The code inside this function strongly resembles the code in `wavelength_theta_figure`.
Can we generalize and reduce duplication?

Reply:
Maybe, but generalizing it might make the logic even more difficult to follow, so I opted to duplicate it here instead.